pkg/agent/monitoraddnodes.go

```diff
@@ -0,0 +1,461 @@
+package agent
+
+import (
+	"context"
+	"io"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	gatherssh "github.com/openshift/installer/pkg/gather/ssh"
+)
+
+type addNodeState int
+
+// The node states that we observe and report.
+const (
+	InitialState addNodeState = iota
+	WaitingForInitialBoot
+	NodeBootedAndSSHIsUp
+	WaitingForAssistedService
+	AssistedServiceIsUp
+	AssistedServiceFailedToStart // terminal state
+	ClusterIsImported
+	ClusterImportedFailed // terminal state
+	InfraenvIsRegistered
+	InfraenvRegisterFailed // terminal state
+	HostConfigsApplied
+	HostConfigsApplyFailed // terminal state
+	WaitingForHostToBeReady
+	HostHasErrors // terminal state, host does not pass validation or has some other error
+	HostInstallationStarted
+	CoreosInstallerWritingToDisk
+	HostPreparingForFirstReboot
+	AfterFirstReboot
+	AfterSecondReboot
+	AfterSecondRebootKubeletIsUp
+	FirstCSRPendingApproval
+	SecondCSRPendingApproval
+	NodeJoinedClusterStatusNotReady
+	Finish // terminal state
+)
```

Comment:
By reviewing this part, I'm not fully convinced that all of these fine-grained states were really required, especially considering that part shared with the `Cluster` type about monitoring the Assisted Service activity.
Similarly to the wait-for case, if assisted service is not initialized within a certain amount of time, then it means that either the host didn't boot or there was some other issue. We already developed a method to check that `agentRestAPILive := czero.API.Rest.IsRestAPILive()`, and it would be nice to reuse it to avoid logic duplication.
Similarly, all the other steps (up to `FirstCSRPendingApproval`) are not reported in the standard wait-for, and I think it would be better to avoid making too many exceptions here.
Considering that after the first node reboot the assisted-service is gone, and so there are no ways to monitor what's happening, I'd recommend ijust to check the kubelet status (either by a direct `systemctl status kubelet` ont he host` or even better with a non-intrusive check, like for example verifying the `healtz` probe with a curl if possible).
So this could be potentially simplified to a linear workflow composed by mainly three state: initial AS activity -> Kubelet status -> Appeareance of the CSRs -> NodeJoined

Reply:
Presently, the "wait-for bootstrap" information is has gaps and is sometimes difficult to understand why a cluster installation had failed unless one takes an agent-gather. My goal here was to bridge some of that gap with these finer grain checks.

But I see a potential issue with this approach, where the ssh public key is required to run the nodeCommandOutputContains commands. The add-nodes image contains the ssh private key extracted from the Cluster's install-config. We don't know who will create the add-nodes image and there is no guarantee that the person who is adding the day2 node has access to the matching ssh public key. This leads me to believe that we can't use ssh.

I will find less intrusive means to check the state as you suggested. I think most of current checks will need to be abandoned.